pkg/trustedresources/verify.go

```diff
@@ -64,11 +64,13 @@ type VerificationResult struct {
 }
 
 // VerifyTask verifies the signature and public key against task.
-// Skip the verification when no policies are found and trusted-resources-verification-no-match-policy is set to ignore or warn
-// Return an error when no policies are found and trusted-resources-verification-no-match-policy is set to fail,
-// or the resource fails to pass matched enforce verification policy
+// VerificationResult is returned with different types for different cases:
+// 1) Return VerificationResult with VerificationSkip type, when no policies are found and no-match-policy is set to ignore
+// 2) Return VerificationResult with VerificationPass type when verification passed;
+// 3) Return VerificationResult with VerificationWarn type, when no matching policies and feature flag "no-match-policy" is "warn", or only Warn mode verification policies fail. Err field if filled with the warning;
+// 4) Return VerificationResult with VerificationError type when no policies are found and no-match-policy is set to fail, the resource fails to pass matched enforce verification policy, or there are errors during verification. Err is filled with the err.
 // refSource contains the source information of the task.
-func VerifyTask(ctx context.Context, taskObj *v1beta1.Task, k8s kubernetes.Interface, refSource *v1beta1.RefSource, verificationpolicies []*v1alpha1.VerificationPolicy) error {
+func VerifyTask(ctx context.Context, taskObj *v1beta1.Task, k8s kubernetes.Interface, refSource *v1beta1.RefSource, verificationpolicies []*v1alpha1.VerificationPolicy) *VerificationResult {
```

Comment:
why return a pointer to a VerificationResult instead of a VerificationResult? That way, we don't have to worry about nil checks/nil pointer dereferencing, and what it means if VerifyTask returns nil. Same for VerifyPipeline.

Reply:
Oh I think it's because my previous code return both VerificationResult and error, and it could be nil. I will update this to not return a pointer